src/library/scala/collection/mutable/TreeMap.scala

```diff
@@ -0,0 +1,166 @@
+package scala
+package collection
+package mutable
+
+import scala.collection.generic._
+import scala.collection.mutable.{RedBlackTree => RB}
+
+/**
+ * $factoryInfo
+ *
+ * @define Coll mutable.TreeMap
+ * @define coll mutable tree map
+ */
+object TreeMap extends MutableSortedMapFactory[TreeMap] {
+
+  def empty[A, B](implicit ord: Ordering[A]) = new TreeMap[A, B]()(ord)
+
+  /** $sortedMapCanBuildFromInfo */
+  implicit def canBuildFrom[A, B](implicit ord: Ordering[A]): CanBuildFrom[Coll, (A, B), TreeMap[A, B]] =
+    new SortedMapCanBuildFrom[A, B]
+}
+
+/**
+ * A mutable sorted map implemented using a mutable red-black tree as underlying data structure.
+ *
+ * @param ordering the implicit ordering used to compare objects of type `A`.
+ * @tparam A the type of the keys contained in this tree map.
+ * @tparam B the type of the values associated with the keys.
+ *
+ * @author Rui Gonçalves
+ * @version 2.12
+ * @since 2.12
+ *
+ * @define Coll mutable.TreeMap
+ * @define coll mutable tree map
+ */
+@SerialVersionUID(-2558985573956740112L)
+sealed class TreeMap[A, B] private (tree: RB.Tree[A, B])(implicit val ordering: Ordering[A])
+  extends AbstractSortedMap[A, B]
+  with SortedMap[A, B]
+  with MapLike[A, B, TreeMap[A, B]]
+  with SortedMapLike[A, B, TreeMap[A, B]]
+  with Serializable {
+
+  /**
+   * Creates an empty `TreeMap`.
+   * @param ord the implicit ordering used to compare objects of type `A`.
+   * @return an empty `TreeMap`.
+   */
+  def this()(implicit ord: Ordering[A]) = this(RB.Tree.empty)(ord)
+
+  override def empty = TreeMap.empty
+  override protected[this] def newBuilder = TreeMap.newBuilder[A, B]
+
+  def rangeImpl(from: Option[A], until: Option[A]): TreeMap[A, B] = new TreeMapView(from, until)
+
+  def -=(key: A): this.type = { RB.delete(tree, key); this }
+  def +=(kv: (A, B)): this.type = { RB.insert(tree, kv._1, kv._2); this }
+
+  def get(key: A) = RB.get(tree, key)
+
+  def iterator = RB.iterator(tree)
+  def iteratorFrom(start: A) = RB.iterator(tree, Some(start))
+  def keysIteratorFrom(start: A) = RB.keysIterator(tree, Some(start))
+  def valuesIteratorFrom(start: A) = RB.valuesIterator(tree, Some(start))
+
+  override def size = tree.size
+  override def isEmpty = RB.isEmpty(tree)
+  override def contains(key: A) = RB.contains(tree, key)
+
+  override def head = RB.min(tree).get
+  override def headOption = RB.min(tree)
+  override def last = RB.max(tree).get
+  override def lastOption = RB.max(tree)
+
+  override def keysIterator = RB.keysIterator(tree)
+  override def valuesIterator = RB.valuesIterator(tree)
+
+  override def foreach[U](f: ((A, B)) => U) = RB.foreach(tree, f)
+  override def transform(f: (A, B) => B) = { RB.transform(tree, f); this }
+  override def clear() = tree.root = null
```

Comment:
`: Unit` would be clearer here also (esp. for people who still think in chained `=`).